lib/Sema/TypeCheckDecl.cpp

```diff
@@ -2437,6 +2498,368 @@ static void checkAccessControl(TypeChecker &TC, const Decl *D) {
+        const VarDecl *theVar = NP->getDecl();
+        if (theVar->getFormalAccess() != AccessLevel::Internal)
+          return;
+        if (!theVar->isUsableFromInline())
+          return;
+        if (seenVars.count(theVar) || theVar->isInvalid())
+          return;
+
+        checkTypeAccess(TC, TypeLoc::withoutLoc(theVar->getType()),
+                        theVar, /*checkUsableFromInline=*/true,
+                        [&](AccessScope typeAccessScope,
+                            const TypeRepr *complainRepr,
+                            DowngradeToWarning downgradeToWarning) {
+          TC.diagnose(P->getLoc(),
+                      diag::pattern_type_not_usable_from_inline_inferred,
+                      theVar->isLet(),
+                      isTypeContext,
+                      theVar->getType());
+        });
+        return;
+      }
+
+      auto *TP = dyn_cast<TypedPattern>(P);
+      if (!TP)
+        return;
+
+      // FIXME: We need an access level to check against, so we pull one out of
+      // some random VarDecl in the pattern. They're all going to be the same,
+      // but still, ick.
+      const VarDecl *anyVar = nullptr;
+      TP->forEachVariable([&](VarDecl *V) {
+        seenVars.insert(V);
+        anyVar = V;
+      });
+      if (!anyVar)
+        return;
+      if (anyVar->getFormalAccess() != AccessLevel::Internal)
+        return;
+      if (!anyVar->isUsableFromInline())
+        return;
+
+      checkTypeAccess(TC, TP->getTypeLoc(), anyVar,
+                      /*checkUsableFromInline=*/true,
+                      [&](AccessScope typeAccessScope,
+                          const TypeRepr *complainRepr,
+                          DowngradeToWarning downgradeToWarning) {
+        auto diag = TC.diagnose(P->getLoc(),
+                                diag::pattern_type_not_usable_from_inline,
+                                anyVar->isLet(),
+                                isTypeContext);
+        highlightOffendingType(TC, diag, complainRepr);
+      });
+    });
+    return;
+  }
+
+  case DeclKind::TypeAlias: {
+    auto TAD = cast<TypeAliasDecl>(D);
+
+    checkTypeAccess(TC, TAD->getUnderlyingTypeLoc(), TAD,
+                    /*checkUsableFromInline=*/true,
+                    [&](AccessScope typeAccessScope,
+                        const TypeRepr *complainRepr,
+                        DowngradeToWarning downgradeToWarning) {
+      auto diag =
+        TC.diagnose(TAD,
+                    diag::type_alias_underlying_type_not_usable_from_inline);
+      highlightOffendingType(TC, diag, complainRepr);
+    });
+
+    return;
+  }
+
+  case DeclKind::AssociatedType: {
+    auto assocType = cast<AssociatedTypeDecl>(D);
+
+    // This must stay in sync with diag::associated_type_not_usable_from_inline.
+    enum {
+      ACEK_DefaultDefinition = 0,
+      ACEK_Requirement
+    };
+
+    std::for_each(assocType->getInherited().begin(),
+                  assocType->getInherited().end(),
```

Comment:
This needs to check requirements, not just the inheritance clause, right? (Also, `llvm::for_each`.)

Reply:
That's a good point. As you correctly noted this is copy and pasted from checkAccessControl() so probably the existing code was wrong too.